opnode/rollup/sync/start.go

```diff
@@ -0,0 +1,157 @@
+// Unsafe Block: The highest L2 block. If the L1 Attributes is ahead of the L1 head, it is assumed to be valid,
+// if not, it walks back until it finds the first L2 block whose L1 Origin is canonical in the L1 chain.
+// Safe Block: The highest L2 block whose sequence window has not changed during a reorg.
+func FindL2Heads(ctx context.Context, start eth.L2BlockRef, seqWindowSize uint64,
+	l1 L1Chain, l2 L2Chain, genesis *rollup.Genesis) (unsafe eth.L2BlockRef, safe eth.L2BlockRef, err error) {
+	reorgDepth := 0
+	var prevL1OriginHash common.Hash
+	// First check if the L1 Origin of the start block is ahead of the current L1 head
+	// If so, we assume that this should be the next unsafe head for the sequencing window
+	// We still need to walk back the safe head because we don't know where the reorg started.
+	l1Head, err := l1.L1HeadBlockRef(ctx)
+	if err != nil {
+		return eth.L2BlockRef{}, eth.L2BlockRef{}, err
+	}
+	l2Ahead := start.L1Origin.Number > l1Head.Number
+	var latest eth.L2BlockRef
+
+	// Walk L2 chain until we find the "latest" L2 block. This the first L2 block whose L1 Origin is canonical.
+	for n := start; ; {
+		// Check if l1Origin is canonical when we get to a new epoch
+		if prevL1OriginHash != n.L1Origin.Hash {
+			if ok, err := isCanonical(ctx, l1, n.L1Origin); err != nil {
+				return eth.L2BlockRef{}, eth.L2BlockRef{}, err
+			} else if ok {
+				latest = n
+				break
+			}
+			prevL1OriginHash = n.L1Origin.Hash
+		}
+		// Don't walk past genesis. If we were at the L2 genesis, but could not find the L1 genesis
+		// pointed to from it, we are on the wrong L1 chain.
+		if n.Hash == genesis.L2.Hash || n.Number == genesis.L2.Number {
+			return eth.L2BlockRef{}, eth.L2BlockRef{}, WrongChainErr
+		}
+		// Pull L2 parent for next iteration
+		n, err = l2.L2BlockRefByHash(ctx, n.ParentHash)
+		if err != nil {
+			return eth.L2BlockRef{}, eth.L2BlockRef{}, fmt.Errorf("failed to fetch L2 block by hash %v: %w", n.ParentHash, err)
+		}
+		reorgDepth++
+		if reorgDepth >= MaxReorgDepth {
+			return eth.L2BlockRef{}, eth.L2BlockRef{}, TooDeepReorgErr
+		}
+	}
+	depth := uint64(1) // SeqWindowSize is a length, but we are counting elements in the window.
+	prevL1OriginHash = latest.L1Origin.Hash
+	// Walk from the latest block back 1 Sequence Window of L1 Origins to determine the safe L2 block.
+	for n := latest; ; {
+		// Advance depth if new origin
+		if n.L1Origin.Hash != prevL1OriginHash {
+			depth++
+			prevL1OriginHash = n.L1Origin.Hash
+		}
+		// Walked sufficiently far
+		if depth == seqWindowSize {
+			if l2Ahead {
+				return start, n, nil
+			} else {
+				return latest, n, nil
+			}
+
+		}
+		// Genesis is always safe.
+		if n.Hash == genesis.L2.Hash || n.Number == genesis.L2.Number {
+			safe = eth.L2BlockRef{Hash: genesis.L2.Hash, Number: genesis.L2.Number, Time: genesis.L2Time, L1Origin: genesis.L1}
+			if l2Ahead {
+				return start, safe, nil
+			} else {
+				return latest, safe, nil
+			}
```

Comment:
A bit dubious about this. What about the scenario where `l2ahead == true` (L1 origin of `start` is ahead of know L1 head) and there is a block known to be non-canonical block in the L2 chain?

Example: L1 head is at block `X`.  `start` has an L1 origin with number `X + 1` but its parent L2 block has an L1 origin with number `X` whose blockhash is not canonical. In this case, the algorithm still returns `start` as the unsafe L1 head, which is incorrect.